File: kontiva-android/app/src/main/java/ch/kontiva/android/persistence/EncryptedStore.kt
package ch.kontiva.android.persistence

import ch.kontiva.android.security.Crypto
import ch.kontiva.android.security.KeyVault
import ch.kontiva.android.security.WrappedKeyMaterial
import ch.kontiva.android.security.WrongPassphraseException
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/** On-disk vault locations (mirrors iOS StoreLocation): keystore.json + store.kenc. */
class StoreLocation(baseDir: File) {
    val dir: File = File(baseDir, "Kontiva").apply { mkdirs() }
    val keystore: File = File(dir, "keystore.json")
    val data: File = File(dir, "store.kenc")
}

class StoreLockedException : Exception("vault is locked")
class VaultExistsException : Exception("vault already exists")
class CorruptVaultException : Exception("vault corrupt")

/**
 * The encrypted local store — a faithful port of iOS `EncryptedStore`. The whole
 * dataset is AES-256-GCM sealed under the master key before being written; the
 * master key is itself only ever stored wrapped (keystore, derived from the
 * passphrase). No plaintext private data is ever written to disk. The decrypted
 * key + dataset live here only while unlocked and are dropped on [lock].
 *
 * The expensive KDF runs off the main thread (callers use Dispatchers.Default).
 */
class EncryptedStore(private val loc: StoreLocation) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private var masterKey: ByteArray? = null
    private var current: AppDataset? = null

    /** Whether a vault already exists on disk (→ show Unlock, not Setup). */
    fun hasExistingVault(): Boolean = loc.keystore.exists()

    val isUnlocked: Boolean get() = masterKey != null

    /** First-run setup: generate + wrap a master key and write the initial dataset. */
    fun createVault(passphrase: String, initial: AppDataset = AppDataset.empty) {
        if (hasExistingVault()) throw VaultExistsException()
        val created = KeyVault.create(passphrase)
        loc.keystore.writeText(json.encodeToString(created.material))
        writeDataset(initial, created.masterKey)
        masterKey = created.masterKey
        current = initial
    }

    /** Unlock an existing vault. Throws WrongPassphraseException on a bad passphrase. */
    fun unlock(passphrase: String) {
        val material = readKeystore()
        val key = KeyVault.unwrap(passphrase, material)
        val dataset = readDataset(key)
        masterKey = key
        current = dataset
    }

    /** Drop the decrypted key + dataset from memory. */
    fun lock() {
        masterKey = null
        current = null
    }

    /** Snapshot of the decrypted dataset (only while unlocked). */
    fun snapshot(): AppDataset = current ?: throw StoreLockedException()

    /** Mutate the dataset and persist it (re-encrypt + write). */
    fun mutate(block: (AppDataset) -> AppDataset) {
        val key = masterKey ?: throw StoreLockedException()
        val updated = block(current ?: throw StoreLockedException())
        writeDataset(updated, key)
        current = updated
    }

    /** Re-wrap the same master key under a new passphrase. */
    fun changePassphrase(old: String, new: String) {
        val rotated = KeyVault.changePassphrase(old, new, readKeystore())
        loc.keystore.writeText(json.encodeToString(rotated))
    }

    /** Danger zone: remove all local data. */
    fun deleteAllData() {
        masterKey = null
        current = null
        loc.data.delete()
        loc.keystore.delete()
    }

    // Portable encrypted backup (separate backup passphrase) ------------------

    /** Produce a `.kontivabackup` blob protected by a separate backup passphrase. */
    fun makeBackup(backupPassphrase: String, appVersion: String): ByteArray {
        val dataset = current ?: throw StoreLockedException()
        val payload = BackupPayload(AppDataset.CURRENT_SCHEMA_VERSION, dataset)
        val plaintext = json.encodeToString(payload).toByteArray(Charsets.UTF_8)
        val salt = Crypto.randomBytes(32)
        val iterations = KeyVault.DEFAULT_ITERATIONS
        val backupKey = Crypto.pbkdf2Sha256(Crypto.normalizedPasswordBytes(backupPassphrase), salt, iterations, 32)
        val sealed = Crypto.seal(plaintext, backupKey)
        val container = BackupContainer(
            BACKUP_FORMAT, KeyVault.KDF_ID, KeyVault.CIPHER_ID, salt.b64(), iterations,
            System.currentTimeMillis(), appVersion, dataset.counts(), sealed.b64(),
        )
        return json.encodeToString(container).toByteArray(Charsets.UTF_8)
    }

    /** Validate a backup and preview its contents without changing local data. */
    fun previewBackup(data: ByteArray, backupPassphrase: String): BackupPreview {
        val container = decodeContainer(data)
        openPayload(container, backupPassphrase) // validates the passphrase
        return BackupPreview(container.createdAt, container.appVersion, container.counts)
    }

    /** Guarded restore: replace ALL local data with the backup contents. */
    fun restoreBackup(data: ByteArray, backupPassphrase: String) {
        val key = masterKey ?: throw StoreLockedException()
        val container = decodeContainer(data)
        val payload = openPayload(container, backupPassphrase)
        writeDataset(payload.dataset, key)
        current = payload.dataset
    }

    private fun decodeContainer(data: ByteArray): BackupContainer {
        val c = try {
            json.decodeFromString<BackupContainer>(String(data, Charsets.UTF_8))
        } catch (_: Exception) {
            throw CorruptVaultException()
        }
        if (c.format != BACKUP_FORMAT) throw CorruptVaultException()
        return c
    }

    private fun openPayload(container: BackupContainer, passphrase: String): BackupPayload {
        val backupKey = Crypto.pbkdf2Sha256(Crypto.normalizedPasswordBytes(passphrase), container.salt.unb64(), container.iterations, 32)
        val plaintext = try {
            Crypto.open(container.sealed.unb64(), backupKey)
        } catch (_: Exception) {
            throw WrongPassphraseException()
        }
        return try {
            json.decodeFromString<BackupPayload>(String(plaintext, Charsets.UTF_8))
        } catch (_: Exception) {
            throw CorruptVaultException()
        }
    }

    // I/O ---------------------------------------------------------------------

    private fun readKeystore(): WrappedKeyMaterial =
        try {
            json.decodeFromString(loc.keystore.readText())
        } catch (_: Exception) {
            throw CorruptVaultException()
        }

    private fun writeDataset(dataset: AppDataset, key: ByteArray) {
        val sealed = Crypto.seal(json.encodeToString(dataset).toByteArray(Charsets.UTF_8), key)
        val tmp = File(loc.data.parentFile, loc.data.name + ".tmp")
        tmp.writeBytes(sealed)
        if (!tmp.renameTo(loc.data)) {
            loc.data.writeBytes(sealed)
            tmp.delete()
        }
    }

    private fun readDataset(key: ByteArray): AppDataset {
        if (!loc.data.exists()) return AppDataset.empty
        val plain = Crypto.open(loc.data.readBytes(), key) // throws on tamper
        return try {
            json.decodeFromString(String(plain, Charsets.UTF_8))
        } catch (_: Exception) {
            throw CorruptVaultException()
        }
    }
}

private const val BACKUP_FORMAT = "kontiva.backup.v1"

@kotlinx.serialization.Serializable
data class BackupPayload(val schemaVersion: Int, val dataset: AppDataset)

@kotlinx.serialization.Serializable
data class BackupContainer(
    val format: String,
    val kdf: String,
    val cipher: String,
    val salt: String,
    val iterations: Int,
    val createdAt: Long,
    val appVersion: String,
    val counts: Map<String, Int>,
    val sealed: String,
)

data class BackupPreview(val createdAt: Long, val appVersion: String, val counts: Map<String, Int>)

private fun ByteArray.b64(): String = java.util.Base64.getEncoder().encodeToString(this)
private fun String.unb64(): ByteArray = java.util.Base64.getDecoder().decode(this)

File: KontivaKit/Sources/KontivaPersistence/EncryptedStore.swift
import Foundation
import CryptoKit
import KontivaCore
import KontivaSecurity

/// The encrypted local store. The entire dataset is sealed with AES-256-GCM under
/// the master key before being written; the master key is itself only ever stored
/// wrapped (in the keystore, derived from the passphrase). **No plaintext private
/// data is ever written to disk.**
///
/// Implemented as an `actor` so the (deliberately expensive) KDF and all file I/O
/// run off the main thread and mutable state is race-free. The decrypted master
/// key and dataset live here only while unlocked and are dropped on `lock()`.
public actor EncryptedStore {
    private let location: StoreLocation
    private var masterKey: SymmetricKey?
    private var current: AppDataset?

    public init(location: StoreLocation) {
        self.location = location
    }

    /// Whether a vault already exists on disk (→ show Unlock, not Setup).
    /// Nonisolated: only reads the immutable location and the filesystem.
    public nonisolated func hasExistingVault() -> Bool {
        FileManager.default.fileExists(atPath: location.keystoreURL.path)
    }

    public var isUnlocked: Bool { masterKey != nil }

    // MARK: - Lifecycle

    /// First-run setup: generate a master key, wrap it under the passphrase, and
    /// write an empty encrypted dataset. Throws if a vault already exists.
    public func createVault(passphrase: String,
                            iterations: Int = KeyVault.defaultIterations) throws {
        guard !hasExistingVault() else { throw StoreError.vaultAlreadyExists }
        let (material, key) = try KeyVault.create(passphrase: passphrase, iterations: iterations)
        try writeKeystore(material)
        let dataset = AppDataset.empty
        try writeDataset(dataset, key: key)
        masterKey = key
        current = dataset
    }

    /// Unlock an existing vault. Throws `.wrongPassphrase` on a bad passphrase.
    public func unlock(passphrase: String) throws {
        let material = try readKeystore()
        let key = try KeyVault.unwrap(passphrase: passphrase, material: material)
        let dataset = try readDataset(key: key)
        masterKey = key
        current = dataset
    }

    /// Drop the decrypted key and dataset from memory.
    public func lock() {
        masterKey = nil
        current = nil
    }

    // MARK: - Access & mutation

    /// Snapshot of the current decrypted dataset (only while unlocked).
    public func snapshot() throws -> AppDataset {
        guard let current else { throw StoreError.locked }
        return current
    }

    /// Mutate the dataset and persist it (re-encrypt + atomic write).
    public func mutate(_ block: @Sendable (inout AppDataset) -> Void) throws {
        guard let key = masterKey, var data = current else { throw StoreError.locked }
        block(&data)
        try writeDataset(data, key: key)
        current = data
    }

    /// Replace the whole dataset and persist it.
    public func replace(with dataset: AppDataset) throws {
        guard let key = masterKey else { throw StoreError.locked }
        try writeDataset(dataset, key: key)
        current = dataset
    }

    /// Re-wrap the same master key under a new passphrase. Existing encrypted data
    /// stays readable. Throws `.wrongPassphrase` if the old passphrase is wrong.
    public func changePassphrase(old: String, new: String,
                                 newIterations: Int = KeyVault.defaultIterations) throws {
        let material = try readKeystore()
        let rotated = try KeyVault.changePassphrase(old: old, new: new,
                                                    material: material,
                                                    newIterations: newIterations)
        try writeKeystore(rotated)
    }

    /// Danger zone: remove all local data (keystore + dataset).
    public func deleteAllData() throws {
        masterKey = nil
        current = nil
        try? FileManager.default.removeItem(at: location.dataURL)
        try? FileManager.default.removeItem(at: location.keystoreURL)
    }

    // MARK: - Portable encrypted backup & guarded restore
    //
    // A backup re-encrypts the entire dataset under a key derived from a SEPARATE
    // backup passphrase, so it is portable to another Mac and independent of the
    // local master key. Restore decrypts the backup and re-encrypts the dataset
    // under the *local* master key.

    private static let backupFormat = "kontiva.backup.v1"

    /// Produce a `.kontivabackup` blob protected by `backupPassphrase`.
    public func makeBackup(backupPassphrase: String, appVersion: String) throws -> Data {
        guard let key = masterKey, let dataset = current else { throw StoreError.locked }

        let payload = BackupPayload(schemaVersion: AppDataset.currentSchemaVersion,
                                    dataset: dataset)
        let plaintext = try JSONEncoder().encode(payload)

        let salt = KeyVault.newSalt()
        let iterations = KeyVault.defaultIterations
        let backupKey = KeyVault.deriveKey(passphrase: backupPassphrase, salt: salt, iterations: iterations)
        let sealed = try SecretBox.seal(plaintext, with: backupKey)

        let container = BackupContainer(
            format: Self.backupFormat, kdf: KeyVault.kdfIdentifier, cipher: KeyVault.cipherIdentifier,
            salt: salt, iterations: iterations, createdAt: Date(), appVersion: appVersion,
            counts: dataset.counts, sealed: sealed)
        return try JSONEncoder().encode(container)
    }

    /// Validate a backup and return its preview without changing any local data.
    /// Throws `.wrongPassphrase` if the passphrase doesn't open it, `.corrupted`
    /// if the file is not a valid backup.
    public func previewBackup(data: Data, backupPassphrase: String) throws -> BackupPreview {
        let container = try decodeContainer(data)
        _ = try openPayload(container, passphrase: backupPassphrase) // validates passphrase
        return BackupPreview(createdAt: container.createdAt, appVersion: container.appVersion,
                             counts: container.counts)
    }

    /// Guarded restore: replace ALL local data with the backup contents. The
    /// caller is responsible for confirming this destructive action first.
    public func restoreBackup(data: Data, backupPassphrase: String) throws {
        guard let key = masterKey else { throw StoreError.locked }
        let container = try decodeContainer(data)
        let payload = try openPayload(container, passphrase: backupPassphrase)

        // Replace the dataset (re-encrypted under the local master key).
        try writeDataset(payload.dataset, key: key)
        current = payload.dataset
    }

    private func decodeContainer(_ data: Data) throws -> BackupContainer {
        guard let container = try? JSONDecoder().decode(BackupContainer.self, from: data),
              container.format == Self.backupFormat else {
            throw StoreError.corrupted
        }
        return container
    }

    private func openPayload(_ container: BackupContainer, passphrase: String) throws -> BackupPayload {
        let backupKey = KeyVault.deriveKey(passphrase: passphrase, salt: container.salt,
                                           iterations: container.iterations)
        let plaintext: Data
        do {
            plaintext = try SecretBox.open(container.sealed, with: backupKey)
        } catch {
            throw SecurityError.wrongPassphrase
        }
        guard let payload = try? JSONDecoder().decode(BackupPayload.self, from: plaintext) else {
            throw StoreError.corrupted
        }
        return payload
    }

    // MARK: - I/O

    private func writeKeystore(_ material: WrappedKeyMaterial) throws {
        let data = try JSONEncoder().encode(material)
        try data.write(to: location.keystoreURL, options: [.atomic])
    }

    private func readKeystore() throws -> WrappedKeyMaterial {
        guard let data = try? Data(contentsOf: location.keystoreURL) else { throw StoreError.noVault }
        guard let material = try? JSONDecoder().decode(WrappedKeyMaterial.self, from: data) else {
            throw StoreError.corrupted
        }
        return material
    }

    private func writeDataset(_ dataset: AppDataset, key: SymmetricKey) throws {
        let plaintext = try JSONEncoder().encode(dataset)
        let sealed = try SecretBox.seal(plaintext, with: key)
        try sealed.write(to: location.dataURL, options: [.atomic])
    }

    private func readDataset(key: SymmetricKey) throws -> AppDataset {
        // Vault exists but no data file yet → treat as empty.
        guard let sealed = try? Data(contentsOf: location.dataURL) else { return AppDataset.empty }
        let plaintext = try SecretBox.open(sealed, with: key) // throws .decryptionFailed on tamper
        guard let dataset = try? JSONDecoder().decode(AppDataset.self, from: plaintext) else {
            throw StoreError.corrupted
        }
        return dataset
    }
}
